sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,687 @@
+
+            if hasattr(init, '__array__'):
+                init -= X_mean
+
+        centers_dict = {
+            0: {'data': X, 'weights': sample_weight, 'sse': None,
+                'centroid': None}
+        }
+
+        init_org = init
+        last_center_id = 0
+
+        for n_iter in range(self.n_clusters - 1):
+
+            biggest, _ = max(centers_dict.items(), key=lambda x: x[1]['sse'])
+
+            # If init array is provided -
+            # Take only part of init that is dedicated for that part of bisect
+            if hasattr(init, '__array__'):
+                init = init_org[n_iter: n_iter + 2].copy()
+                # ^ Not sure if that assigning of init points is
+                # here a good idea?
+
+            # Perform Bisection
+            centers, labels = self._bisect(centers_dict[biggest]['data'],
+                                           init,
+                                           centers_dict[biggest]['weights'],
+                                           random_state)
+
+            # Check SSE (Sum of Squared Errors) of each of computed centroids.
+            # SSE is calculated with distances between data points
+            # and assigned centroids
+            errors = self._compute_bisect_errors(centers_dict[biggest]['data'],
+                                                 centers, labels)
+
+            lower_sse_index = 0 if errors[0] < errors[1] else 1
+            higher_sse_index = 1 if lower_sse_index == 0 else 0
+
+            higher_labels = (labels == higher_sse_index)
+            lower_labels = (labels == lower_sse_index)
+
+            # Add both centroids to dict
+            centers_dict[last_center_id + 1] = {
+                'data': centers_dict[biggest]['data'][lower_labels],
+                'weights': centers_dict[biggest]['weights'][lower_labels],
+                'sse': errors[lower_sse_index],
+                'centroid': centers[lower_sse_index]
+            }
+
+            centers_dict[last_center_id + 2] = {
+                'data': centers_dict[biggest]['data'][higher_labels],
+                'weights': centers_dict[biggest]['weights'][higher_labels],
+                'sse': errors[higher_sse_index],
+                'centroid': centers[higher_sse_index]
+            }
+
+            if self.verbose:
+                print(f"Centroid Found: {centers[lower_sse_index]}")
+                print(f"Centroid Found: {centers[higher_sse_index]}")
+
+            last_center_id += 2
+
+            # Delete split cluster from dict
+            del centers_dict[biggest]
+
+        # Extract calculated centroids to array
+        centers = [center[1]['centroid'] for center in centers_dict.items()]
+
+        centers = np.asarray(centers)
+
+        # Restore Original Data
+        if not sp.issparse(X):
+            X += X_mean
+            centers += X_mean
+
+        self.n_iter_ = n_iter + 1
+
+        return centers
+
+    def _bisect_largest_cluster(self, X, init,
```

Comment:
These three methods seem very redundant, it would be nice to try to factorize them into one method.

Reply:
They're not as much redundant as it seems for first glance. But I'll see what can be done

Reply:
At least `_bisect_largest_cluster` and `_bisect_largest_sse` seem extremely redundant. You only need to change the definition of `biggest`, and to avoid computing the error if you don't need it.

Actually, `_bisect_by_biggest_child_sse` could also use the same code with `biggest = last_center_id`.